WEGA/contrib/cpm/cpm.1
.TH CPM 1 WEGA
.ad
.fi
.SH NAME
.IP
cpm \- Lesen und Schreiben von Disketten mit CP/M-Format
.SH SYNTAX
\fBcpm\fR [\fBoption\fR] [\fBfilename\fR]
.SH BESCHREIBUNG
.B cpm
liest und schreibt Dateien mit der Struktur des
CP/M-Filesystems.
.br
Das angegebene File muss ein Floppy-Disk-Laufwerk sein
(/dev/fd* bzw. /dev/rfd*) oder ein Unix-Standard-File
mit der gleichen Struktur. Da es unbequem und langsam ist,
auf das Geraet ueber Special-File zuzugreifen, ist es immer
guenstig eine Kopie des Disketteninhaltes im 
Unix-Standard-Filesystem anzufertigen. Dazu kann das
dd-Kommando benutzt werden.
.PP
.RS
dd if=/dev/rfd1 of=
.I ihrfile
bs=512 count=..
.RE
.PP
Zu den Geraeten /dev/fd* bzw. /dev/rfd*
darf nur von der console aus zugegriffen werden.
.SH FLAGS:
.RS
.IP -d 16
Anzeige des Directory auf Standard-Ausgabe
.IP -B 16
die Dateien die mit Hilfe des c- bzw. C-Flags uebertragen werden
sollen, sind Binaercode - andernfalls Textdateien.
.IP -c\ \fIname1\fR\ \fIname2\fR 16
Kopiere die CP/M-Datei \fIname1\fR in das UNIX-File \fIname2\fR
.IP -C\ \fIname1\fR\ \fIname2\fR 16
Kopiere das Unix-File \fIname1\fR in die CP/M-Datei \fIname2\fR
.IP -p\ \fIname\fR 16
Kopiere die angegebene CP/M-Datei nach Standard-Ausgabe.
.IP -i 16
Uebergang in den interaktiven Modus. (alle o.g. Flags
sind wirkungslos)
.IP -I 16
Initialisiere das angegebene File im CP/M-Systemformat; 
d.h. alle CP/M-Dateien in diesem System werden geloescht.
.IP -s\ \fIn\fR 16
Sektor-Interleaving-Faktor = \fIn\fR; Standard ist 6
.IP -b\ \fIn\fR 16
Blocklaenge = \fIn\fR (in Byte); Standard ist 2K (2048) Byte
.IP -m\ \fIn\fR 16
maximale Anzahl von Directory-Eintraegen = \fIn\fR; Standard ist 128
.IP -l\ \fIn\fR 16
Sektorlaenge (in Bytes) = \fIn\fR; Standard ist 512
.IP -r\ \fIn\fR 16
Anzahl der Sektoren pro Spur = \fIn\fR; Standard ist 18
.IP -S\ \fIn\fR 16
Anzahl der Systemspuren; Standard ist 0
.IP -t\ \fIn\fR 16
Anzahl der Spuren der Diskette; Standard ist 80
.IP
.RE
Wird das i-Flag gesetzt, muss trotzdem der Filename mit angegeben
werden. Existiert das angegebene File nicht, so wird nach entsprechender
Rueckfrage ein CP/M-Filesystem mit diesem Namen im Unixsystem initialisiert.
.br
Die Flags -C, -c und -p schliessen einander aus. 
.SH
KOMMANDOS IM INTERAKTIVEN MODUS
.RS
.IP ccopyin\ \fIunixfile\fR\ \fIcpmfile\fR 30
Kopiere Unix-Binary-File ins CP/M
.IP ccopyout\ \fIcpmfile\fR\ \fIunixfile\fR 30 
Kopiere CP/M-Binary-Datei ins Unix
.IP copyin\ \fIunixfile\fR\ \fIcpmfile\fR 30
Kopiere Unix-Text-File ins CP/M
.IP copyout\ \fIcpmfile\fR\ \fIunixfile\fR 30
Kopiere CP/M-Text-Datei ins Unix
.IP del[ete]\ \fIfilename\fR 30
Synonym fuer erase
.IP dir[ectory] 30
Anzeige Directory des CP/M-Filesystems
.IP dum[p]\ \fIfilename\fR 30
hexadezimale Anzeige der angegebenen Datei auf dem Terminal
.IP era[se]\ \fIfilename\fR 30
Loeschen der angegeben Datei aus dem CP/M-Filesystem
.IP exi[t] 30
Ende; Fortsetzung im Shell
.IP hel[p] 30
Anzeige des help-Textes
.IP log[out] 30
Synonym fuer exit
.IP ls 30
Synonym fuer dir
.IP ren[ame]\ \fIfile1\fR\ \fIfile2\fR 30
Umbenennung der CP/M-Datei \fIfile1\fR in \fIfile2\fR
.IP typ[e]\ \fIfilename\fR 30
Anzeige der genannten CP/M-Datei am Terminal
.IP !\ \fIUNIX-kommando\fR 30
Ausfuehren des UNIX-Kommandos
.IP disk 30
Anzeige der gesetzten Diskettenparameter
.RE
.IP
Die Kommandonamen koennen mit 3 Buchstaben abgekuerzt werden.
CP/M-Dateinamen werden automatisch in Grossbuchstaben umgewandelt.
Die Kopierkommandi' erlauben nicht das Ueberschreiben bereits
existierender Dateien.
.IP
Wird waehrend der Uebertragung das CP/M-Floppy-Geraet voll, wird
die Datei geschlossen und das Kommando abgebrochen. Die bereits
geschriebenen Daten im CP/M-File bleiben erhalten.
.IP
Das Kommando copyout setzt voraus, dass die CP/M-Textdateien als
Zeilenabschluss \fIcr + lf\fR haben und ignoriert \fIcr\fR bei der Uebertragung
ins Unix-File.
copyin fuegt vor jeden \fIlf\fR im Unix-File bei der Uebertragung in die
CP/M-Datei \fIcr\fR ein und schliesst eine Datei mit ^z ab.
.IP
Die binaeren Kopierkommandos fuehren die Uebertragung mit 'raw'-Dateien
durch, womit auch Code-Dateien von und zu Disketten uebertragen
werden koennen.
.IP
Interrupts werden im interaktiven Modus erkannt. Die
Fortsetzung erfolgt mit Eingabeanforderung fuer eine neue
Kommandozeile.
.SH
FILES
.IP
/dev/fd*   bzw.   /dev/rfd*
.br
/usr/new/lib/cpm.hlp
.SH
FEHLERQUELLEN
.IP
CP/M-Nutzernummern werden ignoriert; Dateien, die auf ein
CP/M-Floppy-Filesystem geschrieben werden, erhalten immer die Nutzernummer 0.
.IP
CP/M-Dateinamenerweiterungen laenger als 3 Zeichen werden entsprechend
abgeschnitten.
.IP
Wildcardzeichen sind nicht erlaubt.
.IP
CP/M-Dateinamen werden nicht erkannt, wenn in Zeichen des Namens das
7. Bit eingtragen (POWER-Kommando "SET") bzw. jenes Bit in
Zeichen der Filenamenergaenzung gesetzt ist (POWER-Kommando "SETRO"
und "SETSYS").


WEGA/contrib/cpm/cpm3/cpm.c
/*	cpm.c	1.15	83/06/20	*/
/*
	Copyright (c) 1982, 1983 by
	Helge Skrivervik, UCB.
	All rights reserved.
	Permission granted for use by UNIX* licencees.
	UNIX is a trademark of Bell Labs.

*/

#include <stdio.h>
#include "cpmio.h"
#include "cpmfio.h"

#define BIG	32767

C_FILE	c_iob[C_NFILE];
int	fid;


int	dflag, cflag, iflag, tflag;
int	blksiz	=	2048;	/* default block size */
int	tracks	=	80;	/* default number of tracks */
int	maxdir	=	128;	/* default number of directory slots per disk */
int	seclth	=	512;	/* default sector length [bytes] */
int	sectrk	=	18;	/* default number of sectors per track */
int	skew	=	1;	/* default sector skew factor */
int	n_systrk=	0;	/* default # of system tracks */

int	*bitmap = 0;
int	flp_fmt = 2;
struct directory *dirbuf;

char *string;

main(argc, argv)
	int	argc;
	char	*argv[];
{

	char *cpmname, *unixname, *malloc();
	int xflag, stat=0, Cflag, Iflag, Bflag;
	long flp_size;

	if (argc == 1)
		usage();
	for (; argc > 1 && argv[1][0] == '-'; argc--, argv++) {
		switch (argv[1][1]) {

		case 0:
			break;

		case 'B':
			Bflag++;
			continue;

		case 'd':
			dflag++;
			continue;

		case 'c':
			if (Cflag)
				stat++;
			cpmname = argv[2];
			unixname = argv[3];
			argc -= 2;
			argv += 2;
			cflag++;
			continue;

		case 'i':
			if (isatty(0) && isatty(1)) iflag++;
			continue;

		case 'p':
			cpmname = argv[2];
			++argv;
			--argc;
			tflag++;
			continue;

		case 's':
			string = argv[2];
			++argv;
			skew = number(BIG);
			argc--;
			if ((skew < 1) || (skew > 10)) {
				printf("skew factor out of range\n");
				exit(1);
			}
			continue;

		case 'S':
			string = argv[2];
			++argv;
			n_systrk = number( BIG );
			argc--;
			if((n_systrk < 0) || (n_systrk > 10 )) {
				printf("%d system tracks are impossible\n",
					n_systrk );
				exit(1);
			}
			continue;

		case 't':
			string = argv[2];
			++argv;
			tracks = number( BIG );
			argc--;
			if((tracks > 1048) || (tracks < 36)) {
				printf("%d tracks are impossible\n",
					tracks );
				exit(1);
			}
			continue;


		case 'b':
			string = argv[2];
			++argv;
			blksiz = number(BIG);
			argc--;
			if (blksiz & 0xc3) {
				printf("illegal block size: %d\n",blksiz);
				exit(1);
			}
			continue;

		case 'm':
			string = argv[2];
			++argv;
			maxdir = number(BIG);
			argc--;
			if ((maxdir < 64) || (tracks >1024 )) {
				printf("illegal value of m-flag: %d\n",maxdir);
				exit(1);
			}
			continue;

		case 'l':
			string = argv[2];
			++argv;
			seclth = number(BIG);
			argc--;
			if ((seclth < 128) || (tracks >1024 )) {
				printf("illegal sector length: %d\n",seclth);
				exit(1);
			}
			continue;

		case 'r':
			string = argv[2];
			++argv;
			sectrk = number(BIG);
			argc--;
			if (sectrk > 48) {
				printf("illegal r-flag: %d\n",sectrk);
				exit(1);
			}
			continue;

		case 'C':
			if (cflag)
				stat++;
			cpmname = argv[3];
			unixname = argv[2];
			argc -= 2;
			argv += 2;
			Cflag++;
			continue;

		case 'I':
			Iflag++;
			continue;

#ifdef DEBUG
		case 'x':
			xflag++;
			continue;
#endif

		default:
			printf("Illegal flag: -%c\n",argv[1][1]);
			stat++;
		}
		break;
	}
	if (stat > 0) {
	}
	flp_size = tracks - n_systrk;
	flp_size *= seclth * sectrk;
	flp_size /= blksiz;
	flp_fmt = flp_size > 256 ? 2 : 1;
	if (argc <= 1 && iflag) {
		printf("cpm: can't copy from stdin in iteractive mode\n");
		exit(1);
	} else {
		if (argc <= 1)
			fid = fileno(stdin);
		else {
			int ic;
			if ((fid = open(*++argv,2)) == -1 ) {
				/*
				 * The specified input file does not exist,
				 * does the user want to initialize a new
				 * file?
				 */

				printf("Input file does not exist.\n");
				printf("Initialize new floppy file? (y/n)?");
				if ((ic = getchar()) != 'y' && ic != 'Y')
					exit(1);
				fid = initcpm(*argv);
				ic = getchar();		/* get <eol> */
			} else {
				if (Iflag) {
				     printf("Initialize floppy file? (y/n)?");
				     if ((ic = getchar()) != 'y' && ic != 'Y')
					exit(1);
				     fid = initcpm(*argv);
				     ic = getchar();		/* get <eol> */
				}
			}
		}
	}
	/* allocate memory for the directory buffer */
	if ((dirbuf = (struct directory *) malloc(maxdir*32)) == NULL ) {
		printf("can't allocate memory for directory\n");
		exit(1);
	}
	gen_sktab();
	getdir();
	build_bmap();
#ifdef DEBUG
	if (xflag > 0) {
		int i;
		char ch;

		dbmap("current bitmap:\n"); 
		for (i = (int)dirbuf; i< (int)dirbuf+maxdir*32; i++) {
			ch = *(char *)i;
			putchar((int)ch);
		}
	}
#endif
	if (iflag > 0) {
		interact();
		exit(0);
	}
	if (dflag > 0) 
		dispdir();
	if (cflag > 0) {
		copy(cpmname, unixname, Bflag);
		exit(0);
	}
	if (Cflag > 0) {
		pipc(unixname, cpmname, Bflag);
		exit(0);
	}
	if (tflag > 0) {
		copy(cpmname, stdout, 0);
		exit(0);
	}
}


number(big)
{
	register char *cs;
	long n;

	cs = string;
	n = 0;
	while(*cs >= '0' && *cs <= '9')
		n = n*10 + *cs++ - '0';
	for(;;)
	switch(*cs++) {

	case 'k':
		n *= 1024;
		continue;

	case 'w':
		n *= sizeof(int);
		continue;

	case 'b':
		n *= 512;
		continue;

	case '*':
	case 'x':
		string = cs;
		n *= number(BIG);

	case '\0':
		if (n >= big || n < 0) {
			fprintf(stderr,"number: argument %D out of range\n", n);
			exit(1);
		}
		return (n);
	}
}

usage()
{
	printf("Uscge: cpm [-i][-d][-p name][-c|C name1 name2] file-name\n");
	exit(1);
}
